### PR TITLE
feat(environmental_information): implement ADDRESS opcode

### PR DESCRIPTION
**Pull request by lambda-0x**
## Pull Request type

Please check the type of change your PR introduces:

- [x] Feature

## What is the current behavior?

Resolves: #116 

## What is the new behavior?

- implements ADDRESS opcode
- add tests for the same

## Does this introduce a breaking change?

- [x] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes

### CODE REVIEW

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    let expected_address: felt252 = evm_address().into();
+    let expected_address: u256 = expected_address.into();
+    assert(address == expected_address, 'stack top should be 0x00');
+}
+
+#[test]
+#[available_gas(20000000)]
+#[ignore]
+fn test_address_nested_call() {// TODO: Figure out a way to do nested calls
+}
```
Comment (lambda-0x) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo]: is there any way to do nested smart contract calls to check that `ADDRESS` works as expected?

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    let expected_address: felt252 = evm_address().into();
+    let expected_address: u256 = expected_address.into();
+    assert(address == expected_address, 'stack top should be 0x00');
+}
+
+#[test]
+#[available_gas(20000000)]
+#[ignore]
+fn test_address_nested_call() { // TODO: Figure out a way to do nested calls
+}
```
Comment (lambda-0x) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:29]: currently is there any way to simulate a nested smart contract calls to check that `ADDRESS` works as expected?

Comment (Eikix) [reply crates/evm/src/tests/test_instructions/test_environment_information.cairo:28]: no, you'll need to create an issue to track this over time

File: crates/evm/src/instructions/environmental_information.cairo [hunk 3]
```diff
@@ -1,85 +1,127 @@
...
+use evm::context::BoxDynamicExecutionContextDestruct;
+use evm::errors::EVMError;
+
+#[generate_trait]
+impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+    /// 0x30 - ADDRESS 
+    /// Get address of currently executing account.
+    /// # Specification: https://www.evm.codes/#30?fork=shanghai
+    fn exec_address(ref self: ExecutionContext) -> Result<(), EVMError> {
+        let current_addr = self.evm_address();
```
Comment (Eikix) [inline crates/evm/src/instructions/environmental_information.cairo:16]: can you implement Into<EthAddress, u256> somewhere where it makes sense in the repo?
and then directly write:
```let current_address: u256 = self.evm_address().into();```
?

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+fn test_address_basic() {
+    // Given
+    let mut ctx = setup_execution_context();
+
+    // When
+    ctx.exec_address();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    let expected_address: felt252 = evm_address().into();
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:19]: nicely done!

**Review by Eikix — COMMENTED** [review]
overall lgtm

File: crates/evm/src/instructions/environmental_information.cairo (hunk 3 above)
Comment (lambda-0x) [reply crates/evm/src/instructions/environmental_information.cairo:16]: oh right, i forget we can implement foreign traits on foreign types unlike rust :), will do.

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo (hunk 2 above)
Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_environment_information.cairo:28]: created https://github.com/kkrt-labs/kakarot-ssj/issues/183, will link it in comments to the test function.

**Review by Eikix — APPROVED** [review]
lgtm 🔥

**Pull request merged**